streams/src/test/java/org/apache/kafka/streams/kstream/internals/SuppressScenarioTest.java

```diff
@@ -146,6 +160,384 @@ public void shouldImmediatelyEmitEventsWithZeroEmitAfter() {
+            .stream("input", Consumed.with(STRING_SERDE, STRING_SERDE))
+            .groupBy((String k, String v) -> k, Serialized.with(STRING_SERDE, STRING_SERDE))
+            .windowedBy(SessionWindows.with(5L).grace(5L))
+            .count(Materialized.<String, Long, SessionStore<Bytes, byte[]>>as("counts").withCachingDisabled());
+        valueCounts
+            .suppress(untilWindowCloses(unbounded()))
+            .toStream()
+            .map((final Windowed<String> k, final Long v) -> new KeyValue<>(k.toString(), v))
+            .to("output-suppressed", Produced.with(STRING_SERDE, Serdes.Long()));
+        valueCounts
+            .toStream()
+            .map((final Windowed<String> k, final Long v) -> new KeyValue<>(k.toString(), v))
+            .to("output-raw", Produced.with(STRING_SERDE, Serdes.Long()));
+        final Topology topology = builder.build();
+        System.out.println(topology.describe());
+        final Properties config = Utils.mkProperties(Utils.mkMap(
+            Utils.mkEntry(StreamsConfig.APPLICATION_ID_CONFIG, getClass().getSimpleName().toLowerCase(Locale.getDefault())),
+            Utils.mkEntry(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, "bogus")
+        ));
+        final ConsumerRecordFactory<String, String> recordFactory = new ConsumerRecordFactory<>(STRING_SERIALIZER, STRING_SERIALIZER);
+        try (final TopologyTestDriver driver = new TopologyTestDriver(topology, config)) {
+            // first window
+            driver.pipeInput(recordFactory.create("input", "k1", "v1", 0L));
+            driver.pipeInput(recordFactory.create("input", "k1", "v1", 1L));
+            // new window
+            driver.pipeInput(recordFactory.create("input", "k1", "v1", 7L));
+            // late event for first window - this should get dropped from all streams, since the first window is now closed.
+            driver.pipeInput(recordFactory.create("input", "k1", "v1", 1L));
+            // just pushing stream time forward to flush the other events through.
+            driver.pipeInput(recordFactory.create("input", "k1", "v1", 30L));
+            verify(
+                drainProducerRecords(driver, "output-raw", STRING_DESERIALIZER, LONG_DESERIALIZER),
+                asList(
+                    new KeyValueTimestamp<>("[k1@0/0]", 1L, 0L),
+                    new KeyValueTimestamp<>("[k1@0/0]", null, 1L),
+                    new KeyValueTimestamp<>("[k1@0/1]", 2L, 1L),
+                    new KeyValueTimestamp<>("[k1@7/7]", 1L, 7L),
+                    new KeyValueTimestamp<>("[k1@30/30]", 1L, 30L)
+                )
+            );
+            verify(
+                drainProducerRecords(driver, "output-suppressed", STRING_DESERIALIZER, LONG_DESERIALIZER),
+                asList(
+                    // TODO: it's not strictly necessary to emit these in final mode, but it's also not harmful... maybe?
+                    new KeyValueTimestamp<>("[k1@0/0]", null, 1L),
```

Comment:
The todo refers to emitting this `null` result, which is a result of session windows merging.

I suppose the suppression operator could drop results with null values if (and only if) it's in "final results" mode. But on the other hand, I worry that it's too much magic.

WDYT?

Reply:
Here's my thinking: for session windows, when we merge the window we will send `negating` old values downstream, but with final output we should not send any data yet before merging and sending the final result, so there is no point sending `null`, and hence in this case we should remove it in the suppressed result as well?

Reply:
Ok, it's a little more complex to maintain as a special case, but it does seem better not to burden our users with mysteries for the sake of convenience.

Reply:
I agree with dropping `null` values